docs/RFCS/20190619_constraint_conformance_report.md

```diff
@@ -0,0 +1,319 @@
+
+# Summary
+
+The set of features described here aim to provide admins with visibility into
+aspects of replication and replica placement. In particular admins will be able
+to access a report that details which replication constraints are violated.
+
+
+# Motivation
+
+As of this writing, CRDB administrators have poor insight into data placement in
+relation to the constraints they have set and in relation with other
+considerations - specifically replication factor and replica diversity. This can
+lead to real problems: it’s unclear when a whole locality can be turned down
+without losing quorum for any ranges.  
+As a result of this work, admins will also be able to script actions to be taken
+when partitioning, or another zone config change, has been “fully applied”.
+
+
+# Guide-level explanation
+
+## Background
+
+Reminder for a few CRDB concepts:
+
+**Replication zone configs and constraints:** key spans can be grouped into
+*“replication zones” and properties can be set for each zone through a “zone
+*config”. Among others, a zone config specifies the replication factor for the
+*data in the zone (i.e. the replication factor of all the ranges in the zone), a
+*list of constraints, and a list of lease preference policies.  
+**Replication Constraints:** used to restrict the set of nodes/stores that can
+*hold a replica for the respective ranges. Constraints can say that only nodes
+*marked with certain localities or attributes are to be used or that certain
+*nodes/stores have to be avoided. Constraints can also be specified at the level
+*of a replica, not only a range; for example, constraints can be set to require
+*that some ranges have one replica in locality A, one in B and one in C.  
+**Lease preferences:** Lease preferences are expressed just like constraints, except
+they only affect the choice of a leaseholder. For example, one can say that the
+leaseholder needs to always be in a specific locality. Unlike constraints, not
+being able to satisfy a lease preference doesn’t stop a lease acquisition.  
+**Replica diversity:** Given a set of constraints (or no constraints), CRDB tries to
+split replicas evenly between localities/sublocalities that match the
+constraint.  
+
+## Constraint conformance
+
+Currently the only insight an administrator has into the health of replication
+is via the Admin UI which displays a counter for under-replicated ranges and one
+for unavailable ranges.
+There’s nothing about constraints, lease preferences or diversity. Also there’s
+no information on over-replicated ranges - although hopefully those will go away
+once we get atomic group membership changes. Also, the way in which these
+counters are computed is defective: the counters are incoherent and, for
+example, if all the nodes holding a partition go away, the ranges in that
+partition are not counted (see
+https://github.com/cockroachdb/cockroach/issues/19644).
+
+Besides the general need for an administrator to know that everything is
```

Comment:
We also learned from a customer this week that they would like to know if the cluster has experienced corruption. I could imagine something like zero fails of the consistency checker on any range in the last 24 hrs

Reply:
it's not per se--and totally fine if we want to pull it out. It does relate to the notion of, is my cluster in a healthy state before I do X